sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs

```diff
@@ -873,5 +1063,277 @@ public virtual Task FlushInternalAsync(CancellationToken cancellationToken = def
+                throw;
+            }
+            finally
+            {
+                Logger.BufferedProducerBackgroundProcessingStartComplete(Identifier, EventHubName);
+            }
+        }
+
+        /// <summary>
+        ///   Performs the actions needed to stop processing events. the <see cref="EventHubBufferedProducerClient" /> to begin accepting events to be
+        ///   enqueued and published.  Should this method be called while processing is not active, no action is taken.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the stop operation.</param>
+        ///
+        /// <remarks>
+        ///   This method will modify class-level state and should be synchronized.  It is assumed that callers hold responsibility for
+        ///   ensuring synchronization concerns; this method should be invoked after any primitives have been acquired.  Callers are also
+        ///   assumed to own responsibility for any validation that the client is in the intended state before calling.
+        ///
+        ///   This method does not consider any events that have been enqueued and are in a pending state.  It is assumed
+        ///   that the caller has responsibility for invoking <see cref="FlushInternalAsync" /> or <see cref="ClearInternalAsync" />
+        ///   prior to this method if needed.
+        /// </remarks>
+        ///
+        private async Task StopPublishingAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            Logger.BufferedProducerBackgroundProcessingStop(Identifier, EventHubName);
+
+            var capturedExceptions = default(List<Exception>);
+
+            try
+            {
+                // If there is no task running for the background management process, there
+                // is nothing to stop.
+
+                if (_producerManagementTask == null)
+                {
+                    return;
+                }
+
+                // Request cancellation of the background processing.
+
+                _publishingCancellationSource?.Cancel();
+                _publishingCancellationSource?.Dispose();
+                _publishingCancellationSource = new CancellationTokenSource();
+
+                // Wait for the background tasks to complete.
+
+                try
+                {
+                    await _producerManagementTask.ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                    // This is expected; no action is needed.
+                }
+                catch (Exception ex)
+                {
+                    Logger.BufferedProducerBackgroundProcessingStopError(Identifier, EventHubName, ex.Message);
+                    (capturedExceptions ??= new List<Exception>()).Add(ex);
```

Comment:
Could we lose an exception here given the wrong multi-thread timing?

Reply:
This method is intended to be called only when the class state semaphore is held, so there should be no concurrency issues.    I had the assumptions detailed in the remarks; please let me know if they aren't clear in signaling the intent:   _(L1162-1170)_

```csharp
        /// <remarks>
        ///   This method will modify class-level state and should be synchronized.  It is assumed that callers hold responsibility for
        ///   ensuring synchronization concerns; this method should be invoked after any primitives have been acquired.  Callers are also
        ///   assumed to own responsibility for any validation that the client is in the intended state before calling.
        ///
        ///   This method does not consider any events that have been enqueued and are in a pending state.  It is assumed
        ///   that the caller has responsibility for invoking <see cref="FlushInternalAsync" /> or <see cref="ClearInternalAsync" />
        ///   prior to this method if needed.
        /// </remarks>
```